### ISSUE TITLE
Pattern to ignore all arguments to a data constructor?

### ISSUE THREAD

**Issue by hdgarrood** [purescript/purescript#1860]
In Haskell, you can do this:

```haskell
data FooBar = Foo String Int | Bar String Number

isFoo :: FooBar -> Bool
isFoo Foo{} = True
isFoo _ = False
```

And the `Foo{}` pattern will continue to work, even if the number of arguments to the `Foo` data constructor changes in the future. Could such a pattern be possible in PureScript? I guess we'd want  to change the syntax a bit, given that `{}` currently means "empty record" in PS.

**Comment by garyb** [comment]
:+1: although agree about finding an alternate syntax. I always thought `Foo{}` was ugly anyway :wink:

**Comment by LiamGoodacre** [comment]
:+1: suggestion of possible syntax:
~~~purescript
isFoo (Foo ...) = True
isFoo _ = False
~~~

**Comment by jdegoes** [comment]
```
isFoo (Foo ...) = True
```

:+1: Yes please.

**Comment by kRITZCREEK** [comment]
If we were to introduce View Patterns could we introduce syntax that matches this case? Or have syntax to specify a "View pattern that only matches a certain constructor without binding its arguments"?

**Comment by garyb** [comment]
Another :+1: for `(Foo ...)` for the standard pattern syntax at least.

**Comment by rgrempel** [comment]
Another possibility would be double- or triple-underscore ... i.e. a "super-underscore".

```purescript
-- double
isFoo (Foo __) = True
isFoo _ = False

-- triple
isFoo (Foo ___) = True
isFoo _ = False
```

**Comment by paf31** [comment]
I'm not sure I like the idea of adding syntax for this actually. If you have so many data constructor arguments that writing them out becomes a chore, then you probably should be using a record. Plus, I don't like adding lots of little bits of syntax.

**Comment by hdgarrood** [comment]
Ok, that seems reasonable. For me, though, the motivation is not avoiding the chore of writing them out, more avoiding unnecessary coupling. I don't care about how long it takes to write them all out, I just don't want to have to update them whenever the number of arguments changes.

**Comment by garyb** [comment]
That was my thought too, the desire for this ocurred to me when dealing with the markdown AST in SlamData. It's place {} is most useful in psc also.

**Comment by paf31** [comment]
Ah ok, that's a reasonable complaint. Once again, I find myself wishing I'd just implemented polymorphic variants with a single argument :smile:

**Comment by paf31** [comment]
Although, I will say, _both_ problems are solved by using a record.

**Comment by hdgarrood** [comment]
Yeah, true. I think data constructors with multiple arguments have advantages too - performance (at least in V8, because you can use the "class" trick), and also it's immediately clear for people reading the code later if you haven't matched on every argument. Whether these outweigh the advantages of records I couldn't say. Although I guess it doesn't really matter because we're not considering deprecating data constructors with multiple arguments anyway?

**Comment by paf31** [comment]
> Although I guess it doesn't really matter because we're not considering deprecating data constructors with multiple arguments anyway?

I think that would be a bit too disruptive, even for me.

**Comment by jdegoes** [comment]
> Ah ok, that's a reasonable complaint. Once again, I find myself wishing I'd just implemented polymorphic variants with a single argument :smile:

Please, it's not too late!!! :laughing:

**Comment by paf31** [comment]
Move to Backlog?

**Comment by garyb** [comment]
Yeah I think so, if we've convinced you of the utility of them :smile:

**Comment by garyb** [comment]
Is "Backlog" now 1.0.0?

**Comment by paf31** [comment]
Yeah, I wanted to do some work to figure out what things we wanted to do before 1.0, and this seemed like the easiest way. It's not final, just a place where we can move things as we prioritize.
